drivers/sensor/adxl345/adxl345.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2020 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/sensor.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <logging/log.h>
+
+#include "adxl345.h"
+
+LOG_MODULE_REGISTER(ADXL345, CONFIG_SENSOR_LOG_LEVEL);
+
+static int adxl345_read_sample(struct device *dev,
+			       struct adxl345_sample *sample)
+{
+	struct adxl345_dev_data *data = dev->driver_data;
+	s16_t raw_x, raw_y, raw_z;
+	u8_t axis_data[6];
+
+	int rc = i2c_burst_read(data->i2c_master,
+				data->i2c_addr,
+				ADXL345_X_AXIS_DATA_0_REG,
+				axis_data,
+				6);
+
+	if (rc < 0) {
+		LOG_ERR("Samples read failed with rc=%d\n", rc);
+		return rc;
+	}
+
+	raw_x = axis_data[0] | (axis_data[1] << 8);
+	raw_y = axis_data[2] | (axis_data[3] << 8);
+	raw_z = axis_data[4] | (axis_data[5] << 8);
+
+	sample->x = raw_x;
+	sample->y = raw_y;
+	sample->z = raw_z;
+
+	return 0;
+}
+
+static void adxl345_accel_convert(struct sensor_value *val, unsigned int sample)
+{
+	if (sample & BIT(9)) {
+		sample |= ADXL345_COMPLEMENT;
+	}
+
+	val->val1 = (sample * 1000) / 32;
+	val->val2 = 0;
+}
+
+static int adxl345_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct adxl345_dev_data *data = dev->driver_data;
+	struct adxl345_sample sample;
+	u8_t samples_count;
+	int rc;
+
+	data->sample_number = 0;
+	rc = i2c_reg_read_byte(dev, data->i2c_addr, ADXL345_FIFO_STATUS_REG,
+			&samples_count);
```

Comment:
check or assert that `samples_count` is less than `ADXL345_MAX_FIFO_SIZE` so you don't overrun the buffer in the loop below

Reply:
done